keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,313 @@
+- **Non-recursive**: For an array or a map, one can not remove or add new items, but existing items can be modified. 
+  For maps this means that the the key stay the same, but values might change. For arrays this means that the length
+  stays the same.
+- **Recursive with allowed addition and deletion**:
+  - New/deleted keys in maps are tolerated.
+  - ~~Appended items or removal of the last item in arrays are tolerated.~~
+    For arrays an identifier field is required, e.g. `.name`.
+- ~~**Mutable**: A field set as mutable could cancel the recursive immutability.~~ probably we don't want this.
+
+In CRDs these modes could be set via `x-kubernetes-immutable-mode: non-recursive/recursive/...`. Inside
+of Golang types the `+immutable` marker could be extended to `// +immutable=non-recursive/recursive/...`.
+
+More complex modes like allowed addition and removal with a custom identifier can be expressed with a
+multi key-value syntax:
+
+```
+// No single field in existing containers can be changed, added or deleted.
+// New containers can be added, existing containers can be removed. Containers
+// are identified by its name.
+// +immutable=addition=allowed,removal=allowed,identifier=.name
+Containers []Container
+```
+
+### Future outline: protobuf
+
+Protobuf encoding identifies unset, empty and null values for slices and maps. In native types
+we use a semantic equality to factor those differences out. In CRs decoding and encoding
+is always faithful with the traditional JSON wire format.
+
+When adding protobuf encoding to CRDs in the future, we have to (without major, non-standard
+effort) identify unset, empty and null for CRs as well. This leads to the idea to use a less
+strict equality in the immutabiltiy checks with that case in mind. But protobuf encoding in
+native types also has a normalization effect, namely posted JSON object are normalized through
+the encoding to protobuf when writing to etcd. 
+
+Hence, it looks sensible to split normalization from immutability equality, and keep a strict 
+deep-equal equality even for protobuf, and potentially native types (if we decide to implement 
+immutability for those).
+
+With the container example we get this, with a sketch of a `+normalize` marker which
+normalizes empty and null to undefined:
+
+```
+// No single field in existing containers can be changed, added or deleted.
+// New containers can be added, existing containers can be removed. Containers
+// are identified by its name.
+// +immutable
+// +normalize=undefined
+Containers []Container `json:containers,omitempty` `protobuf:"bytes,2,opt,name=containers"
+```
+
+For fields which carry no `omitempty`, we could allow more advanced normalization
+modes which replicate the Golang serialization behaviour. Tooling like openapi-gen
+and the CRD validation could verify that the normalization specification matches 
+that of Golang.
```

Comment:
By matches, we mean that when `+normalize=undefined` is set then `json:omitempty` must also be set?

Reply:
yes, that omitempty and the normalization match. Also it might be a pointer type which preserves null vs. omitempty swallows nulls into undefined.